### ISSUE TITLE
external storage does not support recent SSH ciphers

### ISSUE THREAD

**Issue by lpirl** [owncloud/core#16133]
Sadly, I cannot connect to a hardened SSH server since the corresponding library used [does not support recent ciphers/algorithms](https://github.com/phpseclib/phpseclib/issues/676).

### Steps to reproduce
1. harden your SSH server, e.g. according to [this guide](https://stribika.github.io/2015/01/04/secure-secure-shell.html)
2. add it as an external storage

### Expected behaviour
It just works.

### Actual behaviour
The red square is indicating that OC cannot connect.

### Server configuration

**PHP version:** 5.5.0

**ownCloud version:** 8.0.2

**Updated from an older ownCloud or fresh install:** updated

**Are you using external storage, if yes which one:** sftp

**Are you using encryption:** no

**Are you using an external user-backend:** no

#### ownCloud log (data/owncloud.log)
```
{"reqId":"…","remoteAddr":"…","app":"PHP","message":"No compatible key exchange algorithms found at \/var\/www\/vhosts\/example.org\/httpdocs\/owncloud\/3rdparty\/phpseclib\/phpseclib\/phpseclib\/Net\/SSH2.php#1263","level":3,"time":"2015-05-06T20:26:18+00:00","method":"POST","url":"\/index.php\/apps\/files_external\/ajax\/addMountPoint.php"}
{"reqId":"…","remoteAddr":"…","app":"files_external","message":"Exception: {\"Message\":\"Login failed\",\"Code\":0,\"Trace\":\"
#0 \\\/var\\\/www\\\/vhosts\\\/example.org\\\/httpdocs\\\/owncloud\\\/apps\\\/files_external\\\/lib\\\/sftp.php(97): OC\\\\Files\\\\Storage\\\\SFTP->getConnection()\
#1 \\\/var\\\/www\\\/vhosts\\\/example.org\\\/httpdocs\\\/owncloud\\\/apps\\\/files_external\\\/lib\\\/config.php(457): OC\\\\Files\\\\Storage\\\\SFTP->test(false)\
#2 \\\/var\\\/www\\\/vhosts\\\/example.org\\\/httpdocs\\\/owncloud\\\/apps\\\/files_external\\\/lib\\\/config.php(539): OC_Mount_Config::getBackendStatus('\\\\\\\\OC\\\\\\\\Files\\\\\\\\Stora...', Array, false)\
#3 \\\/var\\\/www\\\/vhosts\\\/example.org\\\/httpdocs\\\/owncloud\\\/apps\\\/files_external\\\/ajax\\\/addMountPoint.php(25): OC_Mount_Config::addMountPoint('SFTP', '\\\\\\\\OC\\\\\\\\Files\\\\\\\\Stora...', Array, 'user', 'all', false)\
#4 \\\/var\\\/www\\\/vhosts\\\/example.org\\\/httpdocs\\\/owncloud\\\/lib\\\/private\\\/route\\\/route.php(135) : runtime-created function(1): require_once('\\\/var\\\/www\\\/vhosts...')\
#5 [internal function]: __lambda_func(Array)\
#6 \\\/var\\\/www\\\/vhosts\\\/example.org\\\/httpdocs\\\/owncloud\\\/lib\\\/private\\\/route\\\/router.php(250): call_user_func('\\\\x00lambda_310', Array)\
#7 \\\/var\\\/www\\\/vhosts\\\/example.org\\\/httpdocs\\\/owncloud\\\/lib\\\/base.php(767): OC\\\\Route\\\\Router->match('\\\/apps\\\/files_ext...')\
#8 \\\/var\\\/www\\\/vhosts\\\/example.org\\\/httpdocs\\\/owncloud\\\/index.php(36): OC::handleRequest()\
```

**Comment by PVince81** [comment]
@bantu @LukasReschke

**Comment by LukasReschke** [comment]
Please avoid hyperbole when creating issues. The wording "vulnerable" here is not really appropriate in this context. Yes, PHPSecLib does not support the newest and fanciest ciphers. No, this will not make your data insecure per definition and make kittens die. Yes, this is suboptimal and should eventually get changed in the future.

That said this is an upstream issue which according to the PHPSecLib maintainer will stay open for quite a while: https://github.com/phpseclib/phpseclib/issues/676#issuecomment-99663836

**Comment by lpirl** [comment]
Whether this issue is _"a weakness which allows an attacker to reduce a system's information assurance"_<sup> [1](http://en.wikipedia.org/wiki/Vulnerability_(computing)) </sup> (vulnerability) probably depends on subjective interpretation and requirements regarding confidentiality.
Hence, whether the data is insecure "per definition" or not, depends on the definition.
There is a higher probability of losing confidentiality using the supported algorithms in comparison to more recent ones (no matter if one regards them as "fancy" or not, no matter of kittens).

However, thanks for agreeing that there is some potential in enhancing OC's security.
I just experienced that users could not connect to my server and wanted to raise awareness for this incompatibility.

**Comment by tflidd** [comment]
I tried the patch from @bantu : https://github.com/phpseclib/phpseclib/pull/714
You only need to replace a single file. It worked for me on OC 8.1.1 and an SSH setting from the same guide as @lpirl metioned in https://github.com/owncloud/core/issues/16133#issue-73739160

**Comment by PVince81** [comment]
Does this work now in 8.2.2 ? No sure if the lib was updated to that specific version.

**Comment by tflidd** [comment]
It works for me in OC 8.2.2. I'm not sure, but I think it was already solved in OC 8.1.5.

**Comment by PVince81** [comment]
Cool, thanks
